### PR TITLE
Various performance improvments

### PR DESCRIPTION
**Pull request by kodieg**

### CODE REVIEW

File: src/embedding.rs [hunk 1]
```diff
@@ -76,7 +81,7 @@ where
                         col.insert(j as usize, col_value);
                     }
                 }
-                col
+                crossbeam::utils::CachePadded::new(col)
```
Comment (piobab) [inline src/embedding.rs:84]: It looks like it doesn't give much boost.

File: src/embedding.rs [hunk 2]
```diff
@@ -17,6 +17,10 @@ use std::sync::Arc;
 /// There can be much more but log the first few.
 const LOGGED_NUMBER_OF_BROKEN_ENTITIES: usize = 20;
 
+// FIXME: CachePadded
+pub type Row = crossbeam::utils::CachePadded<Vec<f32>>;
```
Comment (piobab) [inline src/embedding.rs:21]: Can be private

File: src/embedding.rs [hunk 3]
```diff
@@ -17,6 +17,10 @@ use std::sync::Arc;
...
+pub type Matrix = Vec<Row>;
```
Comment (piobab) [inline src/embedding.rs:22]: Actually inner vector in Vec<Vec<f32>> is column so maybe change Row -> Col and write some comment for clarification.

File: src/persistence.rs [hunk 4]
```diff
@@ -130,8 +145,14 @@ pub mod sparse_matrix {
             }
         }
 
-        fn add_pair_index(&mut self, magic: u64, pos: u32) {
-            self.pair_index.insert(magic, pos);
+        fn get_or_add_pair_index(&mut self, magic: u64, pos: u32) -> u32 {
+            match self.pair_index.entry(magic) {
```
Comment (piobab) [inline src/persistence.rs:149]: Can be simplified `self.pair_index.entry(magic).or_insert(pos)`

File: src/persistence.rs [hunk 5]
```diff
@@ -95,23 +107,26 @@ pub mod sparse_matrix {
...
         }
 
-        fn update_entity_counter(&mut self, val: u32) {
-            self.entity_count = val;
-        }
-
-        fn add_hash_id(&mut self, hash: u64, id: u32) {
-            self.hash_2_id.insert(hash, id);
-            self.id_2_hash.insert(id, hash);
+        fn get_or_add_id_by_hash(&mut self, hash: u64) -> u32 {
```
Comment (piobab) [inline src/persistence.rs:120]: I wanted to split storage (persistence) and logic of sparse matrix construction. I would prefer to move this logic to sparse matrix. If we want to use some other storage such as LMDB then we just provide storage without any logic.

File: src/persistence.rs [hunk 6]
```diff
@@ -130,8 +145,14 @@ pub mod sparse_matrix {
             }
         }
 
-        fn add_pair_index(&mut self, magic: u64, pos: u32) {
-            self.pair_index.insert(magic, pos);
+        fn get_or_add_pair_index(&mut self, magic: u64, pos: u32) -> u32 {
```
Comment (piobab) [inline src/persistence.rs:148]: We can remove `get_pair_index` as `get_or_add_pair_index` does what we need.

File: src/persistence.rs [hunk 7]
```diff
@@ -69,7 +81,7 @@ pub mod sparse_matrix {
         entity_count: u32,
         edge_count: u32,
         hash_2_id: FxHashMap<u64, u32>,
-        id_2_hash: FxHashMap<u32, u64>,
+        id_2_hash: Vec<u64>,
```
Comment (piobab) [inline src/persistence.rs:84]: This is nice one :) It gives a lot of boost!

Comment (piobab) [reply src/persistence.rs:84]: In finish() you can update total struct memory usage.

File: src/embedding.rs [hunk 8]
```diff
@@ -64,9 +68,10 @@ where
         let max_hash = 8 * 1024 * 1024;
         let max_hash_float = max_hash as f32;
 
-        let result: Vec<Vec<f32>> = (0..self.dimension)
+        let result: Matrix = (0..self.dimension)
             .into_par_iter()
             .map(|i| {
+                // threre is extreme number of allocations in here
```
[the remark this answers is not in the document]
Comment (kodieg) [reply src/embedding.rs:74]: actually not true anymore

File: src/persistence.rs (hunk 5 above)
Comment (kodieg) [reply src/persistence.rs:120]: I would consider get_or_add operation as an atomic operation that may be provided by any persistent storage. Some databases (not sure about lmdb) allow to implement this in a single operation (thus perform this operation in one and not two blocking calls).

**Comment by piobab** [comment]
LGTM

**Review by piobab — APPROVED** [review]

**Review by piobab — APPROVED** [review]

**Pull request merged**